### PR TITLE
Initial implementation of a windows docker image

### PR DESCRIPTION
**Pull request by dduportal**
This PR introduces a first implementation for the Windows Docker Image:

- Initial issue with discussion: https://github.com/containous/traefik/issues/3603
- This initial image is based on `microsoft/nanoserver:sac2016` for the sake of a first working version
- It introduces an AppVeyor based build to validate the Windows Docker image (a basic and quick smoke test): since updating https://github.com/docker-library/official-images needs a manual PR, the CI won't slow down the Docker Hub's release process nor the Traefik release process, unless I missed something.
  * Build here on my fork: https://ci.appveyor.com/project/dduportal/traefik-library-image
- Changes to the script `update.sh`:
  * A pass of shellcheck had been done against the `update.sh`
  * The generation of the Windows' Dockerfile is now supported by the `update.sh` script.`
  * Faster execution time: the binary downloads when updating scratch's image are paralelized
  * The script `update.sh` had been run again in the repo to ensure idempotency and consistency: no changes outside the windows directory.

### CODE REVIEW

File: windows/tmpl.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,17 @@
+FROM microsoft/nanoserver:$WINDOWS_VERSION
```
Comment (vdemeester) [inline windows/tmpl.Dockerfile]: Wouldn't `BUILD-ARGS` be better for that case (and not having `tmpl` stuff).

```
ARG version sac2016
FROM microsoft/nanoserver:${version}

#…
```

Same for `$VERSION` :angel:

Comment (dduportal) [reply windows/tmpl.Dockerfile]: I'm not sure that build-arg is supported in https://github.com/docker-library/official-images?

Comment (emilevauge) [reply windows/tmpl.Dockerfile]: Yeah, sadly, I think `BUILD-ARGS ` are not supported, neither on official images nor on Docker Hub.

Comment (vdemeester) [reply windows/tmpl.Dockerfile]: :scream:

File: appveyor.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+version: 1.0.{build}
+
+platform: x64
+shallow_clone: true # Because the git clone takes ages due to its size
+
+image:
+- Visual Studio 2017
```
Comment (emilevauge) [inline appveyor.yml]: Could explain in a comment why we use this image ?

File: update.sh [hunk 3]
```diff
@@ -40,38 +44,46 @@ build_from_scratch() {
...
 }
 
 ## From scratch
 build_from_scratch
 
 ## Alpine
-build_alpine
+build_alternate_platform alpine
+
+## Alpine
```
Comment (emilevauge) [inline update.sh]: s/Alpine/Windows

File: appveyor.yml (hunk 2 above)
Comment (dduportal) [reply appveyor.yml]: 👍 Comment added

File: update.sh (hunk 3 above)
Comment (dduportal) [reply update.sh]: 👍 fixed

File: appveyor.yml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+build_script:
+- ps: docker build -t traefik:windows ./windows/
+
+test_script:
+- ps: >-
+    docker run --name lb -d -p 8080:8080 traefik:windows --api --debug
+
+    docker ps
+
+    Invoke-WebRequest -Uri http://localhost:8080
```
Comment (StefanScherer) [inline appveyor.yml]: Oh that's nice. D4W does the magic to use localhost here. :-)
With just Docker EE installed you had to use the container IP instead.

File: appveyor.yml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+- docker version
+
+build:
+  verbosity: minimal
+
+build_script:
+- ps: docker build -t traefik:windows ./windows/
+
+test_script:
+- ps: >-
```
Comment (StefanScherer) [inline appveyor.yml]: When you use `- ps: |` you can omit the empty lines.

```yaml
- ps: |
    docker run ...
    docker ps
    Invoke-WebRequest ...
```

File: appveyor.yml [hunk 6]
```diff
@@ -0,0 +1,30 @@
+version: 1.0.{build}
+
+platform: x64
+shallow_clone: true # Because the git clone takes ages due to its size
+
+# Cf. https://stefanscherer.github.io/setup-windows-docker-ci-appveyor/
```
[the remark this answers is not in the document]
Comment (dduportal) [reply appveyor.yml]: 👍 it is done (I searched the documentation a bit, and I was stuck on the Ubuntu's docker example. Thanks for getting this one here!)

File: appveyor.yml (hunk 4 above)
Comment (dduportal) [reply appveyor.yml]: 👍 I initially used your `curl` container for windows, but the download time was slowing the build. Tried this and it worked: D4W is already set up in Appveyor which is awesome.

File: appveyor.yml (hunk 5 above)
Comment (dduportal) [reply appveyor.yml]: 👍 fixed

**Comment by dduportal** [comment]
Thanks @StefanScherer @emilevauge @SantoDE for the reviews!

I might need a second pass of review: after discussing with @tianon yesterday,
it looks like that we don't have much options left and we have to commit the windows's binary.

I am currently working on this and capturing the information in a markdown file in the repo for future updates :)

**Comment by dduportal** [comment]
Documentation added in the form of "CONTRIBUTING.md" file to convey the process.

Review needed :)

**Pull request merged**